### PR TITLE
Upload complete new feature file

### PR DESCRIPTION
**Pull request by jack-skerrett-bluefruit**

### CODE REVIEW

File: src/jsonifier.py [hunk 1]
```diff
@@ -1,17 +1,23 @@
...
         self.set_tp_format_feature_file()
+        
 
         
     def set_tp_format_feature_file(self):
         for list_test_case in self.reader.feature_file:
+            if(type(list_test_case) == str):
+                if(list_test_case.split()[0] == "Feature:"):
+                    self.feature_name = list_test_case
+                    continue
```
Comment (peter-trevan-bluefruit) [inline src/jsonifier.py:20]: As far as I can tell, isn't `self.reader.feature_file` always a list? When is it a string? 
I'm sure this works fine, I just don't really understand what it's doing here fully.
In general, I think having to rely on testing types may be a sign that you're class needs simplifying or rethinking. However, as with everything there are occasions where it makes perfect sense, and this maybe one of those.

File: settings.py [hunk 2]
```diff
@@ -5,7 +5,10 @@
 NEW_TEST_CASES = {"Name": "","Project":{"ID": ""}, "TestSteps": {"Items": []}}
 DEFAULT_TEST_STEP = {"ResourceType":"TestStep","Description":""}
 ASSIGN_TO_TEST_PLANS = {"TestPlans":{"Items":[]}}
+ASSIGN_TO_FEATURE = {"LinkedTestPlan": {"ResourceType": "TestPlan","Id": ""}}
 ID_DICT_LIST_ITEM = {"ID": ""}
+CREATE_FEATURE_OR_TEST_PLAN_BODY = {"Name":"","Project":{"ID":""}}
+
```
Comment (edmund-hazledine-bluefruit) [inline settings.py:11]: Not sure I would call these settings. Typically settings.py (and full cap variables) is/are used for constants and it looks like these are likely to change?

Comment (edmund-hazledine-bluefruit) [reply settings.py:11]: Oh I see that they're deepcopied. Maybe just move them out of the settings, do any of them get used in more than one place?

File: src/assigner.py [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+        self.feature = feature
+        self.test_plan = test_plan[0]
+        self.test_case = test_case
+        self.data = self.set_data()
+        self.request_url = self.set_request_url()
+
+    def set_data(self):
+        if(self.feature == None):
+            assign_data = deepcopy(ASSIGN_TO_TEST_PLANS)
+            test_case_id_item = deepcopy(ID_DICT_LIST_ITEM)
```
Comment (edmund-hazledine-bluefruit) [inline src/assigner.py:16]: Deepcopy is pretty slow. https://stackoverflow.com/questions/24756712/deepcopy-is-extremely-slow

Comment (edmund-hazledine-bluefruit) [reply src/assigner.py:16]: Perhaps just have these as part of the class?

File: src/assigner.py [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+class Assigner():
+    def __init__(self, feature=None, test_plan=None, test_case=None):
+        self.feature = feature
+        self.test_plan = test_plan[0]
+        self.test_case = test_case
+        self.data = self.set_data()
+        self.request_url = self.set_request_url()
+
+    def set_data(self):
+        if(self.feature == None):
```
Comment (edmund-hazledine-bluefruit) [inline src/assigner.py:14]: If you're equalling to None in python you can use `not` 
```suggestion
        if not self.feature:
```

File: src/assigner.py [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+            test_case_id_item["ID"] = self.test_plan
+            assign_data["TestPlans"]["Items"].append(test_case_id_item)
+            return assign_data
+        elif(self.test_case == None):
+            assign_data = deepcopy(ASSIGN_TO_FEATURE)
+            assign_data["LinkedTestPlan"]["Id"] = self.test_plan
+            return assign_data
+
+    def set_request_url(self):
+        if(self.feature == None):
```
Comment (edmund-hazledine-bluefruit) [inline src/assigner.py:26]: Minor styling preference, I think Peter will disagree but I wouldn't use parenthesis around `if`s. My thinking is that `for` loops don't have them so to keep consistency with that I wouldn't.

Comment (edmund-hazledine-bluefruit) [reply src/assigner.py:26]: And if you are going to keep them I'd put a space after the `if` `elif` 👾

File: tests/test_assigner.py [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+def test_assigner_creates_the_correct_url_to_post_test_case_assignment_to(assign_test_case_to_plan):
+    expected_url = ["https://bluefruit.tpondemand.com/api/v1/TestCases/12345?format=json&access_token=testaccesstoken"]
+    assert assign_test_case_to_plan.request_url == expected_url
+
+@mock.patch("src.assigner.requests.post")
+def test_test_case_is_linked_to_test_plan(mock_request_post, assign_test_case_to_plan):
+    mock_request_post.return_value = mock.Mock(name="mock_api_response",
+                                                **{"status_code": 200, "json.return_value": {"ResourceType": "TestCase","Id": 12345}})
+    assert assign_test_case_to_plan.link_test_case_to_test_plan() == 200
+
```
Comment (peter-trevan-bluefruit) [inline tests/test_assigner.py:42]: Pytest provides it's own mocking system built in, ideally you shouldn't be using `unittest.mock` in pytest files.
Have a look at Pytest's `monkeypatch` fixture, it's built in by default, you don't need to import anything and it's really awesome. https://docs.pytest.org/en/stable/monkeypatch.html

There are a lot of traps with `unittest.mock` that `monkeypatching` will save you from. `unittest.mock` has legitimately coursed me to accidentally make web-requests from within unit tests without realizing it xD.

File: src/assigner.py [hunk 7]
```diff
@@ -0,0 +1,46 @@
+from settings import ID_DICT_LIST_ITEM, ASSIGN_TO_TEST_PLANS, ASSIGN_TO_FEATURE, TP_URL, ACCESS_TOKEN
+from copy import deepcopy
+import requests
+
+class Assigner():
```
Comment (edmund-hazledine-bluefruit) [inline src/assigner.py:5]: Minor thing in python3 you can just do this:
```suggestion
class Assigner:
```

File: src/reader.py [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+    def set_feature_file(self):
+        with open(self.file_name, "r") as f:
+            test_case = []
+            first_pass = True
+            for line in f:
+                if(line.strip() == ""):
+                    continue
+                elif(line.split()[0] == "Feature:"):
+                    self.feature_file.append(line.strip())
+                elif(line[:8] == "Scenario" and first_pass):
```
Comment (edmund-hazledine-bluefruit) [inline src/reader.py:16]: How come you've got a `line.split()[0]` and a `line[:8]`. Does the split not give you Scenario?

File: src/reader.py [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+    def __init__(self, file_name):
+        self.file_name = file_name
+        self.feature_file = []
+        self.set_feature_file()
+
+    def set_feature_file(self):
+        with open(self.file_name, "r") as f:
+            test_case = []
+            first_pass = True
+            for line in f:
```
Comment (edmund-hazledine-bluefruit) [inline src/reader.py:11]: Could you set `line = line.strip()` here to save you doing every time?
```suggestion
            for line in f:
                line = line.strip()
```

Comment (edmund-hazledine-bluefruit) [reply src/reader.py:11]: You can then do `if not line:` on line 12

**Review by edmund-hazledine-bluefruit — COMMENTED** [review]
Looks good! Just some minor comments, feel free to ignore them all 😃

File: tests/test_reader.py [hunk 10]
```diff
@@ -27,7 +31,7 @@ def test_that_there_is_no_empty_line_between_the_last_step_of_a_test_and_an_exam
     i = -1 
     for line in examples_reader.feature_file:
         if(line == "Examples:"):
-            assert examples_reader.feature_file[i] is not ""
+            assert examples_reader.feature_file[i] != ""
         i += 1
```
Comment (peter-trevan-bluefruit) [inline tests/test_reader.py:35]: You can use [enumerate ](https://www.tutorialspoint.com/enumerate-in-python) here :D 
```python
    for i, line in enumerate(examples_reader.feature_file, -1):
        if(line == "Examples:"): 
            assert examples_reader.feature_file[i] != ""
```

**Review by peter-trevan-bluefruit — APPROVED** [review]
Looks good! There's a lot of really clean code in here. Nice work :D

File: src/creater.py [hunk 11]
```diff
@@ -0,0 +1,22 @@
...
+    def upload_entity(self):
+        response = requests.post(self.request_url, json=self.data)
+        if(response.status_code == 200):
+            id_list = []
+            new_entities = response.json()
+            for entity in new_entities["Items"]:
+                id_list.append(entity["Id"])
+            return id_list
+        else:
+            print("Status Code: ", response.status_code)
```
[the remark this answers is not in the document]
Comment (peter-trevan-bluefruit) [reply src/creater.py:22]: This is a general comment. As this pattern appears all throughout

File: src/reader.py (hunk 8 above)
Comment (jack-skerrett-bluefruit) [reply src/reader.py:16]: I can respond to this one now I think! Some tests start with "Scenario:" some with "Scenario Outline:". This means that [0] will sometimes return "Scenario:" and other times "Scenario", a colon makes all the difference :P

File: tests/test_assigner.py (hunk 6 above)
Comment (jack-skerrett-bluefruit) [reply tests/test_assigner.py:42]: Yeah, I didn't want to use unittest/mock but couldn't find any tutorials that were making sense. Maybe now I understand it better I will spend some time converting.

File: src/jsonifier.py (hunk 1 above)
Comment (jack-skerrett-bluefruit) [reply src/jsonifier.py]: Ah, so this is a little quirk with how I am handling the feature name. Its just a string at the top of a list of lists. I think I will make a note to refactor the feature file to be a dict, will probably be more secure.

**Pull request merged**